src/include/catalog/table_handle.h

```diff
@@ -0,0 +1,212 @@
+  std::shared_ptr<SqlTableRW> pg_class_;
+  std::shared_ptr<SqlTableRW> pg_namespace_;
+  std::shared_ptr<SqlTableRW> pg_tablespace_;
+};
+
+/**
+ * A tablespace handle contains information about all the tables in a database.
+ * It is equivalent to pg_tables in postgres, which is a view.
+ * pg_tables (view):
+ *      schemaname | tablename | tablespace
+ *
+ * pg_class:
+ *      __ptr | oid | relname | relnamespace | reltablespace
+ *
+ * pg_namespace:
+ *      oid | nspname
+ *
+ * pg_tablespace:
+ *	    oid | spcname
+ */
+class TableHandle {
+ public:
+  /**
+   * Construct a table handle. It keeps pointers to the pg_class, pg_namespace, pg_tablespace sql tables.
+   * It uses use these three tables to provide the view of pg_tables.
+   * @param catalog a pointer to catalog
+   * @param nsp_oid the namespace oid which the tables belong to
+   * @param pg_class a pointer to pg_class
+   * @param pg_namespace a pointer to pg_namespace
+   * @param pg_tablespace a pointer to pg_tablespace
+   */
+  TableHandle(Catalog *catalog, namespace_oid_t nsp_oid, std::shared_ptr<SqlTableRW> pg_class,
+              std::shared_ptr<SqlTableRW> pg_namespace, std::shared_ptr<SqlTableRW> pg_tablespace)
+      : catalog_(catalog),
+        nsp_oid_(nsp_oid),
+        pg_class_(std::move(pg_class)),
+        pg_namespace_(std::move(pg_namespace)),
+        pg_tablespace_(std::move(pg_tablespace)) {}
+
+  /**
+   * Get the table oid for a given table name
+   * @param txn the transaction context
+   * @param name the table name
+   * @return table oid
+   */
+  table_oid_t NameToOid(transaction::TransactionContext *txn, const std::string &name);
+
+  /**
+   * Get a table entry for the given table oid. It's essentially equivalent to reading a
+   * row from pg_tables. It has to be executed in a transaction context.
+   *
+   * @param txn the transaction that initiates the read
+   * @param oid the table oid
+   * @return a shared pointer to table entry; NULL if the namespace doesn't exist in
+   * the database
+   */
+  std::shared_ptr<TableEntry> GetTableEntry(transaction::TransactionContext *txn, table_oid_t oid);
+
+  /**
+   * Get a table entry for the given table name. It's essentially equivalent to reading a
+   * row from pg_tables. It has to be executed in a transaction context.
+   *
+   * @param txn the transaction that initiates the read
+   * @param name the name of the table
+   * @return a shared pointer to table entry; NULL if the namespace doesn't exist in
+   * the database
+   */
+  std::shared_ptr<TableEntry> GetTableEntry(transaction::TransactionContext *txn, const std::string &name);
+
+  /**
+   * Create a SqlTable. The namespace of the table is the same as the TableHandle.
+   * @param txn the transaction context
+   * @param name the table name
+   * @param schema the table schema
+   */
+  SqlTableRW *CreateTable(transaction::TransactionContext *txn, const Schema &schema, const std::string &name);
```

Comment:
Sometimes your Create methods return `SqlTableRW*` and sometimes they return `std::shared_ptr<catalog::SqlTableRW>`. Please be consistent.

Reply:
We need to return raw pointers and adjust the tests and shutdown code accordingly.

Reply:
These have been converted to raw pointers, so we can control the lifetime of the tables correctly.